### PR TITLE
Hostname resolver

### PR DESCRIPTION
**Pull request by vietj**
ready for review, please review

### CODE REVIEW

**Comment by vietj** [comment]
@pmlopes @cescoffier or any other Vert.x committer please review

File: src/main/java/io/vertx/core/dns/HostnameResolverOptions.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+package io.vertx.core.dns;
+
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Configuration options for Vert.x hostname resolver.
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/dns/HostnameResolverOptions.java]: Add something like: It is used to revolve IP address only (A record only).

File: src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java [hunk 2]
```diff
@@ -43,8 +43,9 @@
    * Called when a socket is listening. For example, this is called when an http or net server
    * has been created and is listening on a specific host/port.
    *
+   * @param localName
    * @param localAddress the local address the net socket is listening on.
    */
-  void listening(SocketAddress localAddress);
+  void listening(String localName, SocketAddress localAddress);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java]: Breaking change.
Must be documented.

File: src/main/java/io/vertx/core/metrics/impl/DummyVertxMetrics.java [hunk 3]
```diff
@@ -319,7 +319,7 @@ public boolean isEnabled() {
   protected class DummyDatagramMetrics implements DatagramSocketMetrics {
 
     @Override
-    public void listening(SocketAddress localAddress) {
+    public void listening(String localName, SocketAddress localAddress) {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/metrics/impl/DummyVertxMetrics.java]: Breaking change
Must be documented.

File: src/main/java/io/vertx/core/dns/HostnameResolverOptions.java (hunk 1 above)
Comment (vietj) [inline src/main/java/io/vertx/core/dns/HostnameResolverOptions.java]: it does AAAA also, but yes indeed.

File: src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java (hunk 2 above)
Comment (vietj) [inline src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java]: that's SPI, but should be documented anyway

Comment (vietj) [inline src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java]: where is the 3.3 page for this ?

Comment (cescoffier) [inline src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java]: https://github.com/vert-x3/wiki/wiki/3.3.0-Breaking-Changes

Comment (vietj) [inline src/main/java/io/vertx/core/spi/metrics/DatagramSocketMetrics.java]: maybe we should have in the doc a difference between API and SPI ? so it is clear that SPI changes won't affect normal users.

**Pull request merged**